### ISSUE TITLE
Equivalent types are reported incompatible

### ISSUE THREAD

**Issue by dmarcautan** [Microsoft/TypeScript#29839]
**TypeScript Version:** 3.3

**Search Terms:**

* union type
* equivalence

**Code**

The following code fragment results in compilation error for **someFuncGeneric** (*Type 'T1 & T2' is not assignable to type 'Matching<T1 & T2, T1 & T2>'.*) and no errors for **someFuncConcrete**:

```typescript
type Matching<InjectedProps, DecorationTargetProps> = {
    [P in keyof DecorationTargetProps]: P extends keyof InjectedProps
        ? InjectedProps[P] extends DecorationTargetProps[P]
            ? DecorationTargetProps[P]
            : InjectedProps[P]
        : DecorationTargetProps[P];
};

function someFuncGeneric<T1, T2>(t1: T1, t2: T2) {
    const someUnionType: T1 & T2 = { ...t1, ...t2 };
    const matchingSameTypes: Matching<T1 & T2, T1 & T2> = { ...t1, ...t2 };
}

type SomeType1 = { SomeProp1: number };
type SomeType2 = { SomeProp2: number };

function someFuncConcrete(t1: SomeType1, t2: SomeType2) {
    const someUnionType: SomeType1 & SomeType2 = { ...t1, ...t2 };
    const matchingSameTypes: Matching<SomeType1 & SomeType2, SomeType1 & SomeType2> = { ...t1, ...t2 };
}
```

**Comment by ahejlsberg** [comment]
This is working as intended. The relationship holds for the concrete types `SomeType1` and `SomeType2`, but not necessarily for _any_ two types `T1` and `T2` (at least it is beyond the capabilities of the type checker to prove it).

**Comment by dmarcautan** [comment]
@ahejlsberg As the problem is reduced to equation Matching<T1, T1> = T1 couldn't it be proved by substituting T1 by any concrete type (like it's done in **someFuncConcrete**)?
